unused-variable false positive

<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this, make sure that the issue doesn't already exist
  or if it is not closed.

  Is your issue fixed on the preview release?: pip install pylint astroid --pre -U

-->

### Steps to reproduce
This seems to be the minimal reproducer, 
```
def f():
    def g():
        return type

    class C(metaclass=g()):
        pass

    return C
```

### Current behavior
```
test.py:5:4: W0612: Unused variable 'g' (unused-variable)
```

### Expected behavior
No error, since the function is used.

### pylint --version output
```
pylint 2.6.0
astroid 2.4.2
Python 3.8.7 (default, Dec 24 2020, 17:53:09) 
[GCC 10.2.0]
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -247,6 +247,11 @@ Release date: TBA
 
   Closes #6414
 
+* Fix false positive for ``unused-variable`` for classes inside functions
+  and where a metaclass is provided via a call.
+
+  Closes #4020
+
 
 What's New in Pylint 2.13.7?
 ============================

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -607,6 +607,11 @@ Other Changes
 
   Closes #5769
 
+* Fix false positive for ``unused-variable`` for classes inside functions
+  and where a metaclass is provided via a call.
+
+  Closes #4020
+
 * Only raise ``not-callable`` when all the inferred values of a property are not callable.
 
   Closes #5931

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2806,6 +2806,10 @@ class VariablesChecker(BaseChecker):
             while not isinstance(attr, nodes.Name):
                 attr = attr.expr
             name = attr.name
+        elif isinstance(klass._metaclass, nodes.Call) and isinstance(
+            klass._metaclass.func, nodes.Name
+        ):
+            name = klass._metaclass.func.name
         elif metaclass:
             name = metaclass.root().name
 

--- a/tests/functional/b/bugfix_local_scope_metaclass_1177.py
+++ b/tests/functional/b/bugfix_local_scope_metaclass_1177.py
@@ -20,6 +20,16 @@ def func_scope():
     return Class2
 
 
+def func_scope_with_metaclass_from_call():
+    def get_type():
+        return type
+
+    class Class2(metaclass=get_type()):
+        pass
+
+    return Class2
+
+
 class ClassScope:
     class Meta3(type):
         pass

